backend/models/notification.js

```diff
@@ -156,6 +163,81 @@ module.exports = {
 			.then(c => c.deleteMany({}));
 	},
 
+	createUserNotification: async function (users, teamSpace, modelId, issueId) {
+		return Promise.all(users.map(u => {
+			return this.upsertIssueClosedNotification(u, teamSpace, modelId, issueId)
+				.then((n) => {
+					return ({ username: u, notification: n });
+				});
+		}));
+	},
+
+	// BUG :returning authorised users,  but error on promise.
+	allowAccess: async function (users, teamSpace, modelId) {
+		return Promise.all(users.map(user => {
+			return hasWriteAccessToModelHelper(user, teamSpace, modelId)
+				.then((canWrite => {
+					return ({ user, canWrite });
+				}));
+		}));
+	},
+
+	removeUserNotification: async function (users, teamSpace, modelId, issueId, issueType) {
+		return Promise.all(
+			users.map(u => {
+				return this.removeIssueNotification(u, teamSpace, modelId, utils.objectIdToString(issueId), issueType)
+					.then((n) => {
+						return Object.assign({ username: u}, n);
+					});
+			})).then(notifications => notifications.reduce((a, c) => !c.notification ? a : a.concat(c), []));
+
+	},
+
+	findAssignedJobs : async function (issue) {
+		const comments = issue.comments;
+		const rolesKey = "assigned_roles";
+		const assignedRoles = new Set();
+
+		// Add the user who created the issue.
+		assignedRoles.add(issue.creator_role);
+
+		// Add current assigned role.
+		assignedRoles.add(issue.assigned_roles[0]);
+
+		for (const item in comments) {
+			const actionProperty = comments[item].action;
+			// Check for additional roles that have been assigned using the issue comments.
+
+			if (actionProperty && actionProperty.property === rolesKey) {
+				assignedRoles.add(actionProperty.from);
+			}
+		}
+
+		return assignedRoles;
+
+	},
+
+	upsertIssueClosedNotifications: async function (username, teamSpace, modelId, issue) {
+
+		const assignedRoles = await this.findAssignedJobs(issue);
+
+		const matchedUsers = await job.findUsersWithJobs(teamSpace, [...assignedRoles]);
+
+		// Leave out the current user , closing the issue.
+		const users = matchedUsers.filter(m => m !== username);
```

Comment:
this line, L230 and L232 should be merged into one loop. We're doign a lot of unncessary iterations here

Let's say `matchedUsers` has n entries.

L227 - iterate n times, resulting in n-1 items in users.
L230 - `allowAccess` is a map, so you're looping n-1 times, let's say it returns n-1-a (for a = number of users who doesnt' have write Access)
L232 - filter - another loop, for n-1-a.

So the complexity of the whole thing is n(n-1)(n-1-a), which is pretty much n^3. When you can do the whole thing in n.